src/_macosx.m

```diff
@@ -40,60 +40,84 @@
 static bool keyChangeCapsLock = false;
 /* Keep track of the current mouse up/down state for open/closed cursor hand */
 static bool leftMouseGrabbing = false;
-/* Keep track of whether stdin has been received */
-static bool stdin_received = false;
-static bool stdin_sigint = false;
 // Global variable to store the original SIGINT handler
 static PyOS_sighandler_t originalSigintAction = NULL;
 
-// Signal handler for SIGINT, only sets a flag to exit the run loop
+// Stop the current app's run loop, sending an event to ensure it actually stops
+static void stop_with_event() {
+    [NSApp stop: nil];
+    // Post an event to trigger the actual stopping.
+    [NSApp postEvent: [NSEvent otherEventWithType: NSEventTypeApplicationDefined
+                                         location: NSZeroPoint
+                                    modifierFlags: 0
+                                        timestamp: 0
+                                     windowNumber: 0
+                                          context: nil
+                                          subtype: 0
+                                            data1: 0
+                                            data2: 0]
+             atStart: YES];
+}
+
+// Signal handler for SIGINT, only argument matching for stop_with_event
 static void handleSigint(int signal) {
-    stdin_sigint = true;
+    stop_with_event();
+}
+
+// Helper function to flush all events.
+// This is needed in some instances to ensure e.g. that windows are properly closed.
+// It is used in the input hook as well as wrapped in a version callable from Python.
+static void flushEvents() {
+    while (true) {
+        NSEvent* event = [NSApp nextEventMatchingMask: NSEventMaskAny
+                                   untilDate: [NSDate distantPast]
+                                      inMode: NSDefaultRunLoopMode
+                                     dequeue: YES];
+        if (!event) {
+            break;
+        }
+        [NSApp sendEvent:event];
+    }
 }
 
 static int wait_for_stdin() {
-    @autoreleasepool {
-        stdin_received = false;
-        stdin_sigint = false;
+    // Short circuit if no windows are active
+    // Rely on Python's input handling to manage CPU usage
+    // This queries the NSApp, rather than using our FigureWindowCount because that is decremented when events still
+    // need to be processed to properly close the windows.
```

Comment:
This will also make us play nicer with any other OSX apps you launch from Python (which I am not aware of any others....)

Reply:
Should we use this method in the `close()` method too instead of keeping track of it ourselves?

Reply:
There are some folks embedding this in C++/objc it would appear: https://github.com/matplotlib/matplotlib/issues/27147

Reply:
That is its own bit of fun, but I think this will not intersect with that issue because the only place the `PyOS_InputHook` is run is from the Python repl (either readline or the new one) so if they are embedding Python in another application (which means they make an interpreter object and then stuff strings into it to run Python code) so I expect the repl will never run.